perf-testing/test_local_only.rb

```diff
@@ -0,0 +1,267 @@
+    "p50": p50(input),
+    "p90": p90(input),
+    "max": input.last,
+    "n": ITERATIONS
+  }
+  outfile.puts(JSON.pretty_generate(result))
+end
+
+def output_raw(thread, outfile)
+  $stdout = outfile
+  puts 'Json Serde Data Raw'
+  pp thread[:json_serde_data]
+  puts 'Cbor Serde Data Raw'
+  pp thread[:cbor_serde_data]
+  puts 'Request Size Data Raw (Alternating Json and CBOR)'
+  pp thread[:request_size_data]
+  puts 'Response Size Data Raw (Alternating Json and CBOR)'
+  pp thread[:response_size_data]
+  $stdout = STDOUT
+end
+
+def analyze(test_cases, measurements, input, protocol, data)
+  test_cases.each do |test|
+    measurements.each do |m|
+      write_test_output(test, protocol, 0, m, input[test][m], data)
+    end
+  end
+end
+
+def separate_and_analyze(test_cases, measurements, thread, data, raw)
+  raw.puts('All test cases')
+  output_raw(thread, raw)
+
+  json_full_serde_data = Array.new(5) { [] }
+  json_full_request_data = Array.new(5) { [] }
+  json_full_response_data = Array.new(5) { [] }
+  thread[:json_serde_data].each_with_index do |d, i|
+    json_full_serde_data[i % 5] << d
+  end
+  thread[:request_size_data].each_with_index do |d, i|
+    json_full_request_data[i % 5] << d unless i % 10 > 4
+  end
+  thread[:response_size_data].each_with_index do |d, i|
+    json_full_response_data[i % 5] << d unless i % 10 > 4
+  end
+  json_full_test_cases = {}
+  (0...5).each do |i|
+    separated = separate_serde_data(json_full_serde_data[i])
+    current = {}
+    current[measurements[0]] = separated[0]
+    current[measurements[1]] = separated[1]
+    current[measurements[2]] = json_full_request_data[i]
+    current[measurements[3]] = json_full_response_data[i]
+    json_full_test_cases[test_cases[i]] = current
+  end
+  analyze(test_cases, measurements, json_full_test_cases, 'JSON', data)
+
+  cbor_full_serde_data = Array.new(5) { [] }
+  cbor_full_request_data = Array.new(5) { [] }
+  cbor_full_response_data = Array.new(5) { [] }
+  thread[:cbor_serde_data].each_with_index do |d, i|
+    cbor_full_serde_data[i % 5] << d
+  end
+  thread[:request_size_data].each_with_index do |d, i|
+    cbor_full_request_data[i % 5] << d unless i % 10 < 5
+  end
+  thread[:response_size_data].each_with_index do |d, i|
+    cbor_full_response_data[i % 5] << d unless i % 10 < 5
+  end
+  cbor_full_test_cases = {}
+  (0...5).each do |i|
+    separated = separate_serde_data(cbor_full_serde_data[i])
+    current = {}
+    current[measurements[0]] = separated[0]
+    current[measurements[1]] = separated[1]
+    current[measurements[2]] = cbor_full_request_data[i]
+    current[measurements[3]] = cbor_full_response_data[i]
+    cbor_full_test_cases[test_cases[i]] = current
+  end
+  analyze(test_cases, measurements, cbor_full_test_cases, 'CBOR', data)
+end
+
+thread = Thread.current
+thread[:json_serde_data] = []
+thread[:cbor_serde_data] = []
+thread[:request_size_data] = []
+thread[:response_size_data] = []
+thread[:warm] = false
```

Comment:
Rather than having a warm value on Thread and then needing to check it in each handler where we're recording values, you could call `clear_thread_data` after the warm up calls are done.